kernel/arch/dreamcast/include/dc/maple/controller.h

```diff
@@ -24,149 +27,466 @@
-    int joyx;
+/** \brief   Controller buttons for standard reset action
+    \ingroup controller_inputs
+    
+    Convenience macro providing the standard button combination
+    used as a reset mechanism by most retail games.
+*/
+#define CONT_RESET_BUTTONS  (CONT_A | CONT_B | CONT_X | CONT_Y | CONT_START)
 
-    /** \brief  Main joystick y-axis value. */
-    int joyy;
+/** \brief   Controller state structure.
+    \ingroup controller_inputs
 
-    /** \brief  Secondary joystick x-axis value (if applicable). */
-    int joy2x;
+    This structure contains information about the status of the controller
+    device and can be fetched by casting the result of maple_dev_status() to
+    this structure.
 
-    /** \brief  Secondary joystick y-axis value (if applicable). */
-    int joy2y;
-} cont_state_t;
+    A 1 bit in the buttons' bitfield indicates that a button is pressed, and the
+    joyx, joyy, joy2x, joy2 values are all 0 based (0 is centered).
 
-/* \cond */
-/* Init / Shutdown */
-int cont_init(void);
-void cont_shutdown(void);
-/* \endcond */
+    \sa maple_dev_status
+*/
+typedef struct cont_state {
+    union {
+        /** \brief  bit-packed controller button states
+            \sa     controller_buttons
+        */
+        uint32_t buttons;
+        struct {
+            uint32_t c: 1;              /**< \brief C button value. */
+            uint32_t b: 1;              /**< \brief B button value. */
+            uint32_t a: 1;              /**< \brief A button value. */
+            uint32_t start: 1;          /**< \brief Start button value. */
+            uint32_t dpad_up: 1;        /**< \brief Main Dpad Up button value. */
+            uint32_t dpad_down: 1;      /**< \brief Main Dpad Down button value. */
+            uint32_t dpad_left: 1;      /**< \brief Main Dpad Left button value. */
+            uint32_t dpad_right: 1;     /**< \brief Main Dpad Right button value. */
+            uint32_t z: 1;              /**< \brief Z button value. */
+            uint32_t y: 1;              /**< \brief Y button value. */
+            uint32_t x: 1;              /**< \brief X button value. */
+            uint32_t d: 1;              /**< \brief D button value. */
+            uint32_t dpad2_up: 1;       /**< \brief Secondary Dpad Up button value. */
+            uint32_t dpad2_down: 1;     /**< \brief Secondary Dpad Down button value. */
+            uint32_t dpad2_left: 1;     /**< \brief Secondary Dpad Left button value. */
+            uint32_t dpad2_right: 1;    /**< \brief Secondary Dpad Right button value. */
+            uint32_t: 16;
+        };
+    };
+
+    int ltrig;    /**< \brief Left trigger value (0-255). */
+    int rtrig;    /**< \brief Right trigger value (0-255). */
+    int joyx;     /**< \brief Main joystick x-axis value. (-128 - 127) */
+    int joyy;     /**< \brief Main joystick y-axis value. */
+    int joy2x;    /**< \brief Secondary joystick x-axis value (if applicable). */
```

Comment:
Aren't *all* the values in cont_state_t only 'if applicable'?

Reply:
...yes, lol. Good point.